Updated to clarify what a checkpoint is and how to create it (#790)
EventStore/documentation: docs/clients/grpc/subscriptions.md

### Subscribing to `$all`

Subscribing to `$all` is similar to subscribing to a single stream. The handler will be called for every event appended after the starting position.

@[code{subscribe-to-all}](@grpc:subscribing_to_stream.py;subscribing-to-streams.js;subscribing-to-streams.ts;subscribing_to_stream/SubscribingToStream.java;subscribing-to-streams/Program.cs;subscribingToStream.go;subscribing_to_stream.rs)

## Subscribing from a specific position

> The previous examples subscribed to the stream from the beginning. This called the handler for every event in the stream and then waited for new events.

Reviewer: ```suggestion
The previous examples subscribed to the stream from the beginning. That subscription invoked the handler for every event in the stream before waiting for new events.
```

Both stream and $all subscriptions accept a starting position if you want to read from a specific point onward. If events already exist at the position you subscribe to, they will be read on the server side and sent to the subscription.

Once caught up, the server will push any new events received on the streams to the client. There is no difference between catching up and live on the client side.

::: warning
The positions provided to the subscriptions are exclusive. You will only receive the next event after the subscribed position.
:::

### Subscribing to a stream

To subscribe to a stream from a specific position, you must provide a *stream position*. This can be `Start`, `End` or a *big int* (unsigned 64 bit integer) position.

The following subscribes to the stream `some-stream` at position `20`, this means that events `21` and onward will be handled:

@[code{subscribe-to-stream-from-position}](@grpc:subscribing_to_stream.py;subscribing-to-streams.js;subscribing-to-streams.ts;subscribing_to_stream/SubscribingToStream.java;subscribing-to-streams/Program.cs;subscribingToStream.go;subscribing_to_stream.rs)

### Subscribing to $all

> Subscribing to the `$all` stream is similar to subscribing to a regular stream. The only difference is how to specify the stream position. For the `$all` stream, you must provide a `Position` structure instead, consisting of two big integers - prepare and commit positions. The `Position` value can be `Start`, `End`, or a `Position` created from a commit and prepare position.

Reviewer: ```suggestion
Subscribing to the `$all` stream is similar to subscribing to a regular stream. The difference is how to specify the starting position. For the `$all` stream, provide a `Position` structure that consists of two big integers: the prepare and commit positions. Use `Start`, `End`, or create a `Position` from specific commit and prepare values.
```

Reviewer: not sure if that's 100% accurate. Please review

### Filtering by event type

If you only want to subscribe to events of a given type, there are two options. You can either use a regular expression or a prefix.

#### Filtering by prefix

If you want to filter by prefix, pass in a `SubscriptionFilterOptions` to the subscription with an `EventTypeFilter.Prefix`.

@[code{event-type-prefix}](@grpc:server_side_filtering.py;server-side-filtering.js;server-side-filtering.ts;server_side_filtering/ServerSideFiltering.java;server-side-filtering/Program.cs;serverSideFiltering.go;server_side_filtering.rs)

This will only subscribe to events with a type that begin with `customer-`.

#### Filtering by regular expression

It might be advantageous to provide a regular expression when you want to subscribe to multiple event types.

@[code{event-type-regex}](@grpc:server_side_filtering.py;server-side-filtering.js;server-side-filtering.ts;server_side_filtering/ServerSideFiltering.java;server-side-filtering/Program.cs;serverSideFiltering.go;server_side_filtering.rs)

This will subscribe to any event that begins with `user` or `company`.

### Filtering by stream name

> If you want to subscribe to a stream only with a given name, you have two options: you can use a regular expression or a prefix.

Reviewer: ```suggestion
To subscribe to a stream by name, choose either a regular expression or a prefix.
```

#### Filtering by prefix

If you want to filter by prefix, pass in a `SubscriptionFilterOptions` to the subscription with an `StreamFilter.Prefix`.

@[code{stream-prefix}](@grpc:server_side_filtering.py;server-side-filtering.js;server-side-filtering.ts;server_side_filtering/ServerSideFiltering.java;server-side-filtering/Program.cs;serverSideFiltering.go;server_side_filtering.rs)

This will only subscribe to all streams with a name that begins with `user-`.

#### Filtering by regular expression

> If you want to subscribe to multiple streams, providing a regular expression might be better.

Reviewer: ```suggestion
To subscribe to multiple streams, use a regular expression.
```
